src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    goto cleanup;
+  }
+
+  // Encode the image in AVIF format
+
+  avifEncoder *encoder = avifEncoderCreate();
+  int quantizerQuality = quality == DEFAULT_QUALITY ? 
+                         DEFAULT_QUANTIZER : quality2Quantizer(quality);
+
+  encoder->minQuantizer = quantizerQuality;
+  encoder->maxQuantizer = quantizerQuality;
+  encoder->minQuantizerAlpha = quantizerQuality;
+  encoder->maxQuantizerAlpha = quantizerQuality;
+  encoder->speed = speed;
+  setEncoderTilesAndThreads(encoder, &rgb);
+
+  result = avifEncoderAddImage(encoder, avifIm, 1, AVIF_ADD_IMAGE_FLAG_SINGLE); //TODO: why 1?
+  if ((failed = isAvifError(result, "Could not encode image"))) {
+    goto cleanup;
+  }
+
+  result = avifEncoderFinish(encoder, &avifOutput);
+  if ((failed = isAvifError(result, "Could not finish encoding"))) {
+    goto cleanup;
+  }
+
+  // Write the AVIF image bytes to the GD ctx.
+
+  gdPutBuf(avifOutput.data, avifOutput.size, outfile);
+
+  cleanup:
+    if (rgb.pixels) {
+      avifRGBImageFreePixels(&rgb);
+    }
+
+    if (encoder) {
+      avifEncoderDestroy(encoder);
+    }
+
+    if (avifOutput.data) {
+      avifRWDataFree(&avifOutput);
+    }
+
+    return failed;
+}
+
+/*****************************************************************************
+ * 
+ *                            HELPER FUNCTIONS
+ * 
+ *****************************************************************************/
+
+/* Convert the quality param we expose to the quantity params used by libavif.
+   The *Quantizer* params values can range from 0 to 63, with 0 = highest quality and 63 = worst.
+   We make the scale 0-100, and we reverse this, so that 0 = worst quality and 100 = highest.
+
+   Values below 0 are set to 0, and values below MAX_QUALITY are set to MAX_QUALITY.
+*/
+
+static int quality2Quantizer(int quality) {
+  int clampedQuality = AVIF_CLAMP(quality, 0, MAX_QUALITY);
+
+  float scaleFactor = (float) AVIF_QUANTIZER_WORST_QUALITY / (float) MAX_QUALITY;
+
+  return round(scaleFactor * (MAX_QUALITY - clampedQuality));
+}
+
+/*
+  From gd_png.c:
+    convert the 7-bit alpha channel to an 8-bit alpha channel.
+    We do a little bit-flipping magic, repeating the MSB
+    as the LSB, to ensure that 0 maps to 0 and
+    127 maps to 255. We also have to invert to match
+    PNG's convention in which 255 is opaque. 
+*/
+
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha) {
+  return originalAlpha == 127 ? 
+        0 : 
+        255 - ((originalAlpha << 1) + (originalAlpha >> 6));
+}
+
+/*
+   As of February 2021, this algorithm reflects the latest research on how many tiles
+   and threads to include for a given image size.
+   This is subject to change as research continues.
+ */
+
+static void setEncoderTilesAndThreads(avifEncoder *encoder, avifRGBImage *rgb) {
+  int imageArea, tiles, tilesLog2;
+  int tileRowsLog2, tileColumnsLog2, maxThreads;
+
+  imageArea = rgb->width * rgb->height;
+
+  tiles = ceil(imageArea / MIN_TILE_AREA);
```

Comment:
Using tiles makes encoding faster, but at the same time the file size increases. So what is a goal of gd here, making encoding fast or produce smaller images?

I guess we want smaller files ...

You should use a tile size of 2048x2048. If the dimension of the image get bigger the additional size for tiled images is negligible. If the image with or height is getting above 4096 use increase the tile size to the max of 3072.

Reply:
gd defaults would be commonly accepted defaults for the particular library which balance the two.  then the `Ex` variants of the APIs allows the user to tweak the knobs.

Ben & i talked about exposing lower library-specific state so users could tweak things even more, but we were going to wait on that until after this lands as it's an issue that applies to all backends (png/webp/jpeg/etc...).

Reply:
Yeah, that's a good question. My concern is that AVIF encoding can get really slow. But I see your point too.

The algorithm here is based on research done by a number of people. @wantehchang, do you want to share your experience here?

Reply:
Brief anecdotal, non-scientific test: I just encoded a 4000x3000 image with the default speed and with quantizer=30. I tried it with 4 tiles, 8 tiles, and then 32 tiles. The number of bytes in the resulting file barely varied - 446993, 447012, and 447210. The time varied more - 52 seconds, 44 seconds, and 38.

I'll keep trying this with other params.

Reply:
I first want to clarify that the tiles we are discussing here are the tiles in AV1. The MIAF standard, which the AVIF image format is based on, supports image grids and also refers to the elements of an image grid as tiles. (libavif calls them "grid cells".) Those are not the tiles we are discussing here.

Regarding the use of tiles in AV1, I can only provide two general guidelines from my colleagues who work on the libaom encoder.

> Avoid small tiles because they are particularly bad for image coding. Tiles improve encoding and decoding speeds when multiple threads are available. However, for image coding, the total tile boundary length affects the compression efficiency because intra prediction can't go across tile boundaries. So the more tiles there are in an image, the worse the compression ratio is. For a given number of tiles, making the tile size close to a square tends to reduce the total tile boundary length inside the image.

> Use no more tiles than the number of threads. Aim for one tile per thread. Using more than one thread inside one tile could be less efficient. Using more tiles than the number of threads would result in a compression penalty without much benefit.

Tiles can help speed up decoding because tiles can be decoded in parallel. So that's another factor to consider when deciding whether tiles should be used.

The 512x512 minimum tile area was suggested by a colleague of mine. You can certainly use a larger minimum, but 2048x2048 seems too big. Alternatively, you can deal with tiles in a follow-up pull request.

Reply:
If no one objects, I'll stick with what we have now - which is based on current research by a number of people. Let me know if we should revisit this.